src/storage-client/src/storage_collections.rs

```diff
@@ -1818,6 +2009,137 @@ where
+                // `compare_and_downgrade_since`.
+                PersistEpoch(None)
+            }
+        }
+    }
+
+    pub async fn compare_and_downgrade_since(
+        &mut self,
+        expected: &PersistEpoch,
+        new: (&PersistEpoch, &Antichain<T>),
+    ) -> Result<Antichain<T>, PersistEpoch> {
+        match self {
+            Self::Critical(handle) => handle.compare_and_downgrade_since(expected, new).await,
+            Self::Leased(handle) => {
+                let (opaque, since) = new;
+                assert!(opaque.0.is_none());
+
+                handle.downgrade_since(since).await;
+
+                Ok(since.clone())
+            }
+        }
+    }
+
+    pub async fn maybe_compare_and_downgrade_since(
+        &mut self,
+        expected: &PersistEpoch,
+        new: (&PersistEpoch, &Antichain<T>),
+    ) -> Option<Result<Antichain<T>, PersistEpoch>> {
+        match self {
+            Self::Critical(handle) => {
+                handle
+                    .maybe_compare_and_downgrade_since(expected, new)
+                    .await
+            }
+            Self::Leased(handle) => {
+                let (opaque, since) = new;
+                assert!(opaque.0.is_none());
+
+                handle.maybe_downgrade_since(since).await;
+
+                Some(Ok(since.clone()))
+            }
+        }
+    }
+
+    pub fn snapshot_stats(
+        &self,
+        id: GlobalId,
+        as_of: Option<Antichain<T>>,
+    ) -> BoxFuture<'static, Result<SnapshotStats, StorageError<T>>> {
+        match self {
+            Self::Critical(handle) => {
+                let res = handle
+                    .snapshot_stats(as_of)
+                    .map(move |x| x.map_err(|_| StorageError::ReadBeforeSince(id)));
+                Box::pin(res)
+            }
+            Self::Leased(handle) => {
+                let res = handle
+                    .snapshot_stats(as_of)
+                    .map(move |x| x.map_err(|_| StorageError::ReadBeforeSince(id)));
+                Box::pin(res)
+            }
+        }
+    }
+
+    pub fn snapshot_stats_from_txn(
+        &self,
+        id: GlobalId,
+        data_snapshot: DataSnapshot<T>,
+    ) -> BoxFuture<'static, Result<SnapshotStats, StorageError<T>>> {
+        match self {
+            Self::Critical(handle) => Box::pin(
+                data_snapshot
+                    .snapshot_stats_from_critical(handle)
+                    .map(move |x| x.map_err(|_| StorageError::ReadBeforeSince(id))),
+            ),
+            Self::Leased(handle) => Box::pin(
+                data_snapshot
+                    .snapshot_stats_from_leased(handle)
+                    .map(move |x| x.map_err(|_| StorageError::ReadBeforeSince(id))),
+            ),
+        }
+    }
+
+    pub async fn expire(self) {
+        match self {
+            Self::Critical(_handle) => {
+                unreachable!("not trying to expire critical handles this way");
```

Comment:
Not sure if this is a recognized idiom or just in my head, but I'd prefer using `unreachable!` only in cases where the unreachability follows from the local context and could be inferred by a smarter compiler. In the case, it follows from how the method is used and the code can quickly become reachable if some remote code changes, so using `panic!` and documenting the panic condition in the doc string seems better.

Reply:
That sounds reasonable, will do!